src/batch/src/executor/monitor/stats.rs

```diff
@@ -11,16 +14,140 @@
+                loop {
+                    match delete_queue_receiver.try_recv() {
+                        Ok(collector) => {
+                            delete_cache.push(collector);
+                        }
+                        Err(tokio::sync::mpsc::error::TryRecvError::Empty) => {
+                            break;
+                        }
+                        Err(_) => {
+                            // Error handle need modify later.
+                            error!("delete_queue_receiver is closed");
+                            connect = false;
+                            break;
+                        }
+                    }
+                }
+            }
+        });
+
+        Self {
+            registry,
+            sender: delete_queue_sender,
+        }
+    }
+
+    pub fn create_task_metrics(&self, id: TaskId) -> BatchTaskMetrics {
+        BatchTaskMetrics::new(self.registry.clone(), id, Some(self.sender.clone()))
+    }
+
+    /// Create a new `BatchTaskMetricsManager` instance used in tests or other places.
+    pub fn unused() -> Self {
+        let (delete_queue_sender, _) = tokio::sync::mpsc::channel::<Box<dyn Collector>>(1);
+        Self {
+            sender: delete_queue_sender,
+            registry: prometheus::Registry::new(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct BatchTaskMetrics {
+    sender: Option<Sender<Box<dyn Collector>>>,
+    pub exchange_recv_row_number: GenericCounterVec<AtomicU64>,
+}
+
+impl BatchTaskMetrics {
+    pub fn new(registry: Registry, id: TaskId, sender: Option<Sender<Box<dyn Collector>>>) -> Self {
+        let opt = {
+            let const_labels = HashMap::from([
+                ("query_id".to_string(), id.query_id),
+                ("target_staget_id".to_string(), id.stage_id.to_string()),
+                ("target_task_id".to_string(), id.task_id.to_string()),
+            ]);
+            opts!(
+                "batch_exchange_recv_row_number",
+                "Total number of row that have been received from upstream source",
+            )
+            .const_labels(const_labels)
+        };
+        let exchange_recv_row_number = register_int_counter_vec_with_registry!(
+            opt,
+            &["source_stage_id", "source_task_id"],
+            registry
+        )
+        .unwrap();
+        Self {
+            sender,
+            exchange_recv_row_number,
+        }
+    }
+
+    /// This function execute after the exucution done.
+    /// Send all the record to the delete queue.
+    pub async fn clear_record(&self) {
+        if let Some(sender) = self.sender.as_ref() {
+            if sender
+                .send(Box::new(self.exchange_recv_row_number.clone()))
+                .await
+                .is_err()
+            {
+                error!("Failed to send delete record to delete queue");
+            }
+        }
+    }
+
+    pub fn unused() -> Self {
```

Comment:
1. Please add some comments to clarify that it's used for tests.
2. Add `#[cfg(test)]` so that it will not be misused by production code.
3. `unused` seems wierd to me, how about `for_test`?

Reply:
I find this function used in #[tokio::test] https://github.com/singularity-data/risingwave/blob/c56c90a99e3846455c94d76089bbd6496680cc38/src/compute/tests/integration_tests.rs#L220
If I add #[cfg(test)], this is invisible in #[tokio::test]. I will get the error like this:
```
error[E0599]: no function or associated item named `for_test` found for struct `risingwave_batch::executor::BatchMetrics` in the current scope
   --> src/compute/tests/integration_tests.rs:278:32
    |
278 |         Arc::new(BatchMetrics::for_test()),
    |                                ^^^^^^^^ function or associated item not found in `risingwave_batch::executor::BatchMetrics`
```
And I can't find the conditions like #[cfg(test)] to make it visible in #[tokio::test].
Do you some way to solve this, otherwise we can just remove #[cfg(test)] and add the comment.

Reply:
Oh, I forget this. Yes we can only add comments in this case.